design/system_tests.md

```diff
@@ -0,0 +1,41 @@
+# System Tests
+
+## Requirements
+1. Test the system in the same manner users will be using it. 
+   No ad-hoc initializations or mocking of any kind. 
+2. Cover the most common use-cases.
+3. Portability: The tests don't assume anything on the environment where the lakeFS & DB are running.
+   lakeFS address is the only **required** argument for running the tests.  
+
+## Non-Requirements
+1. Test every use-case of the system, including:
+    a. Every API/gateway endpoint.
+    b. Every application (e.g. Spark) usage pattern.
+2. Avoid repeating unit/integration tests scenarios.
+3. Stress testing / benchmarks. 
+
+## Solution
+### Test
+A test consists of one or more calls to the API/gateway and assertions on the actual output.
+Test is a standalone unit. It shouldn't rely on any other test to pass to start execution. 
+
+Going forward we can consider adding checks on the DB or data lake layer(S3),
+by adding information of the underlying system components.
+
+#### Test-Cases
+TBD
+
+### Tests binary
+Tests will be built into a single binary and run sequentially or in parallel.
+
+### When will it run? Where?
+As a standalone unit, the tests binary can be executed from anywhere as long as it can reach the lakeFS endpoint.
+We encourage developers to run the tests locally as part of the development cycle. 
+
+#### CI
+Github workflow will be triggered when all other workflows finished successfully and reviewers approved.
+The workflow will run the lakeFS app with local postgres DB on a dedicated server,
+execute the tests binary and report the results. Failed run will block the PR merge.
+
+S3 bucket will be created for every run and deleted when done.
```

Comment:
IME this can be surprisingly hard to do:  In order to debug failures it is really nice to have the written data left over.  So deleting the bucket always is hard.  On the other hand, if you keep buckets, _even only on failures_, they accumulate -- and you need to delete them later.

Maybe we should prefer to be less isolated here?  Straw person idea, more to explain what I mean than as necessarily the right thing to do: All workflows share the same bucket.  Each run gets its own prefix (and is responsible to stay inside it).  Keep all failed or even all runs, and set an S3 expiry on the entire bucket (or some shared prefix).  I think this can give both of my goals.